### PR TITLE
feat: BIP-340 Schnorr Signature 

### PR DESCRIPTION
**Pull request by egeaybars123**
BIP-340 has been implemented in Cairo by the JoyBoy community, and it would be a good addition to add it to Alexandria.

## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/math/src/schnorr.cairo [hunk 1]
```diff
@@ -0,0 +1,127 @@
+//! bip340 implementation
```
Comment (maciejka) [inline packages/math/src/schnorr.cairo:1]: Rename the module back to bip-340, it is not a general purpose schnorr, it uses double sha256 as a hash function which is very bitcoin specific.

**Review by maciejka — CHANGES_REQUESTED** [review]

File: packages/math/src/schnorr.cairo (hunk 1 above)
Comment (egeaybars123) [reply packages/math/src/schnorr.cairo]: Done!

**Review by maciejka — APPROVED** [review]

File: packages/math/src/bip340.cairo [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+/// - `rx`: `u256` - The x-coordinate of the R point from the signature.
+/// - `px`: `u256` - The x-coordinate of the public key.
+/// - `m`: `ByteArray` - The message for which the signature is being verified.
+///
+/// # Returns:
+/// `sha256(tag) || sha256(tag) || bytes(rx) || bytes(px) || m` as u256 where tag =
+/// "BIP0340/challenge".
+fn hash_challenge(rx: u256, px: u256, m: ByteArray) -> u256 {
+    // sha256(tag)
+    let [x0, x1, x2, x3, x4, x5, x6, x7] = compute_sha256_byte_array(@"BIP0340/challenge");
```
[the remark this answers is not in the document]
Comment (egeaybars123) [reply packages/math/src/bip340.cairo:43]: Thanks! I can commit the changes in a few hours.

[the remark this answers is not in the document]
Comment (egeaybars123) [reply packages/math/src/bip340.cairo]: @gaetbout Done!

[the remark this answers is not in the document]
Comment (gaetbout) [reply packages/math/src/bip340.cairo]: Hey, my bad, I should have been more specific...
I believe you can even push it further :)

[the remark this answers is not in the document]
Comment (egeaybars123) [reply packages/math/src/bip340.cairo]: Ahh I see now hahaha. Will start working on it later today. Thanks!

**Comment by gaetbout** [comment]
@0xLucqs can you also have a look to this?
Would appreciate a second pair of eyes 😄

**Comment by egeaybars123** [comment]
image
This was the best I could come up with @gaetbout :D

**Review by gaetbout — APPROVED** [review]

File: packages/math/src/bip340.cairo [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+///
+/// This function checks if the signature `(rx, s)` is valid for a message `m` with
+/// respect to the public key `px`.
+///
+/// # Parameters
+/// - `px`: `u256` - The x-coordinate of the public key.
+/// - `rx`: `u256` - The x-coordinate of the R point from the signature.
+/// - `s`: `u256` - The scalar component of the signature.
+/// - `m`: `ByteArray` - The message for which the signature is being verified.
+///
```
Comment (gaetbout) [inline packages/math/src/bip340.cairo:77]: I don't think that the documentation is following the standard (cfr other files)
Once that is fixed, it'll be merged :)

**Pull request merged**